### PR TITLE
Several improvements that are necessary for v2

### PR DESCRIPTION
**Pull request by mpnowacki-reef**
Solves, amongst others: https://github.com/Backblaze/b2-sdk-python/issues/195

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * `B2RawApi` renamed to `B2RawHTTPApi`
 * `B2HTTP` tests are now common
 * `B2HttpApiConfig` class introduced to provide parameters like `user_agent_append` to `B2Api` without using internal classes in v2
+* `Bucket.update` return s a `Bucket` object in v2
```
Comment (ppolewicz) [inline CHANGELOG.md:15]: extra space

File: b2sdk/bucket.py [hunk 2]
```diff
@@ -115,17 +115,16 @@ def set_type(self, bucket_type):
...
+        lifecycle_rules: Optional[dict] = None,
+        if_revision_is: Optional[int] = None,
         default_server_side_encryption: Optional[EncryptionSetting] = None,
         default_retention: Optional[BucketRetentionSetting] = None,
     ):
         """
         Update various bucket parameters.
-        For legacy reasons in apiver v1 it returns whatever server returned on b2_update_bucket call, v2 will change that.
 
         :param str bucket_type: a bucket type
```
Comment (ppolewicz) [inline b2sdk/bucket.py:129]: maybe worth mentioning allPublic and allPrivate here, since it's not easily available anywhere near here

File: b2sdk/bucket.py [hunk 3]
```diff
@@ -136,16 +135,19 @@ def update(
...
-            account_id,
-            self.id_,
-            bucket_type=bucket_type,
-            bucket_info=bucket_info,
-            cors_rules=cors_rules,
-            lifecycle_rules=lifecycle_rules,
-            if_revision_is=if_revision_is,
-            default_server_side_encryption=default_server_side_encryption,
-            default_retention=default_retention,
+        return BucketFactory.from_api_bucket_dict(
```
Comment (ppolewicz) [inline b2sdk/bucket.py]: by convention it should be `self.BUCKET_FACTORY.from_api_bucket_dict(`

File: test/unit/bucket/test_bucket.py [hunk 4]
```diff
@@ -828,6 +828,110 @@ def _make_file(self, bucket=None):
...
+                'lifecycle_rules': {
+                    'life': 'is life'
+                },
+                'options_set':
+                    set(),
+                'default_server_side_encryption':
+                    SSE_B2_AES,
+                'default_retention':
+                    BucketRetentionSetting(RetentionMode.COMPLIANCE, RetentionPeriod(years=7)),
+            }.items():
```
Comment (ppolewicz) [inline test/unit/bucket/test_bucket.py]: this really needs `# yapf: disable` and manual formatting

File: b2sdk/file_version.py [hunk 5]
```diff
@@ -206,7 +206,7 @@ def __init__(
         content_disposition: Optional[str],
         content_length: int,
         content_language: Optional[str],
-        expires: Optional[datetime],
+        expires: Optional[str],
```
Comment (ppolewicz) [inline b2sdk/file_version.py:215]: why?

File: b2sdk/api.py [hunk 6]
```diff
@@ -440,34 +442,54 @@ def create_key(
             valid_duration_seconds=valid_duration_seconds,
             bucket_id=bucket_id,
             name_prefix=name_prefix
+            # As of 17.06.2021 the server does not accept an 'options' argument
```
Comment (ppolewicz) [inline b2sdk/api.py:445]: that's because it's a read-only property, You can use it to see if the bucket is S3-capable or not (and I guess legalhold-capable nowadays?)

File: b2sdk/v1/api.py [hunk 7]
```diff
@@ -27,6 +27,8 @@
 # and to retain old style download_file_by_id signature (allowing for the new one as well) and exception
 # and to retain old style get_file_info return type
 # and to accept old-style raw_api argument
+# and to retain old style download_file_by_id signature (allowing for the new one as well) and exception
```
Comment (ppolewicz) [inline b2sdk/v1/api.py:30]: this here line 30 is a clone of line 27

File: b2sdk/v1/api.py [hunk 8]
```diff
@@ -169,3 +171,34 @@ def download_file_by_id(
                 raise
         else:
             return downloaded_file
+
+    def list_keys(self, start_application_key_id=None) -> dict:
+        """
+        List application keys.
+
+        :param start_application_key_id: an :term:`application key ID` to start from or ``None`` to start from the beginning
```
Comment (ppolewicz) [inline b2sdk/v1/api.py:179]: it should be noted that only a max of `self.DEFAULT_LIST_KEY_COUNT` is going to be returned (unlike v2 which returns all keys, unless you stop iterating, I guess?)

File: b2sdk/exception.py [hunk 9]
```diff
@@ -485,6 +485,10 @@ def __str__(self):
         return "%s cannot be used as default for a bucket." % (self.encryption_mode,)
 
 
+class CopyArgumentsMismatch(B2Error):
```
Comment (ppolewicz) [inline b2sdk/exception.py:488]: no common abstract exception for user input errors?

File: test/unit/sync/test_sync_report.py [hunk 10]
```diff
@@ -10,12 +10,10 @@
...
 from unittest.mock import MagicMock
 
-from ..test_base import TestBase
+from apiver_deps import SyncReport
 
-from .deps import SyncReport
 
-
-class TestSyncReport(TestBase):
+class TestSyncReport:
```
Comment (ppolewicz) [inline test/unit/sync/test_sync_report.py:18]: it no longer inherits from TestBase!

File: b2sdk/v1/sync/report.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+
+# override to retain legacy methods
+class SyncReport(v2.SyncReport):
+    @property
+    def local_file_count(self):
+        return self.total_count
+
+    @local_file_count.setter
+    def local_file_count(self, value):
```
Comment (ppolewicz) [inline b2sdk/v1/sync/report.py:21]: does this really work? It looks like it would shadow the earlier one

File: b2sdk/bucket.py [hunk 12]
```diff
@@ -135,7 +135,7 @@ def update(
         :param b2sdk.v1.BucketRetentionSetting default_retention: bucket default retention setting
         """
         account_id = self.api.account_info.get_account_id()
-        return BucketFactory.from_api_bucket_dict(
+        return self.api.BUCKET_FACTORY_CLASS.from_api_bucket_dict(
```
Comment (ppolewicz) [inline b2sdk/bucket.py:138]: ❤️

File: b2sdk/api_config.py [hunk 13]
```diff
@@ -8,30 +8,33 @@
...
 ######################################################################
 
+from typing import Optional, Callable, Type
 import requests
 
-from typing import Optional
-
-from types import ModuleType
+if False:
+    from .raw_api import AbstractRawApi
```
Comment (ppolewicz) [inline b2sdk/api_config.py]: ?

File: b2sdk/session.py [hunk 14]
```diff
@@ -67,7 +67,7 @@ def __init__(
         :param api_config
         """
 
-        self.raw_api = B2RawHTTPApi(B2Http(api_config))
+        self.raw_api = (api_config.raw_api_class or B2RawHTTPApi)(B2Http(api_config))
```
Comment (ppolewicz) [inline b2sdk/session.py:70]: I don't think we should be setting a default here. Config must have it.

File: b2sdk/encryption/setting.py [hunk 15]
```diff
@@ -20,7 +21,12 @@
 
 logger = logging.getLogger(__name__)
 
-UNKNOWN_KEY_ID = ...
+
+class _UnknownKeyId(enum.Enum):
+    unknown_key_id = 0
+
+
+UNKNOWN_KEY_ID = _UnknownKeyId.unknown_key_id
```
Comment (ppolewicz) [inline b2sdk/encryption/setting.py:29]: docstring please - what is it

File: b2sdk/encryption/setting.py [hunk 16]
```diff
@@ -20,7 +21,12 @@
 
 logger = logging.getLogger(__name__)
 
-UNKNOWN_KEY_ID = ...
+
+class _UnknownKeyId(enum.Enum):
```
Comment (ppolewicz) [inline b2sdk/encryption/setting.py:25]: docstring/comment please - why is it made this way?

File: b2sdk/api_config.py (hunk 13 above)
Comment (mpnowacki-reef) [reply b2sdk/api_config.py]: The symbol is needed for type annotations but the module cannot be imported because that would cause circular dependencies

File: b2sdk/session.py (hunk 14 above)
Comment (mpnowacki-reef) [reply b2sdk/session.py:70]: Yes. I didn't want to put it explicitly as a default value for `B2HttpApiConfig.__init__`, because `B2RawHttpApi` is part of the internal API, but it should be an evaluated default there. As some say: "there is a better way"

File: b2sdk/api_config.py (hunk 13 above)
Comment (mpnowacki-reef) [reply b2sdk/api_config.py]: I've fixed the circular imports - they were caused by the tests in raw_api.

File: b2sdk/api_config.py [hunk 17]
```diff
@@ -34,7 +33,7 @@ def __init__(
         self.http_session_factory = http_session_factory
         self.install_clock_skew_hook = install_clock_skew_hook
         self.user_agent_append = user_agent_append
-        self.raw_api_class = _raw_api_class
+        self.raw_api_class = _raw_api_class or B2RawHTTPApi
```
Comment (ppolewicz) [inline b2sdk/api_config.py:36]: I think it should be `self.DEFALT_RAW_API_CLASS` for future proofing like we did earlier in this PR

File: b2sdk/file_version.py (hunk 5 above)
Comment (mpnowacki-reef) [reply b2sdk/file_version.py:215]: Parsing to datetime didn't work as it was implemented, and in general I've found it hard to parse these timestamps using builtin python libraries

File: test/unit/sync/test_sync_report.py (hunk 10 above)
Comment (mpnowacki-reef) [reply test/unit/sync/test_sync_report.py:18]: correct. It was moved to the new test structure, which uses pytest and no longer requires test classes to inherit from TestBase. In fact, it sometimes hurts when they do.

File: b2sdk/v1/sync/report.py (hunk 11 above)
Comment (mpnowacki-reef) [reply b2sdk/v1/sync/report.py:21]: what do you mean? The getter and the setter? they don't work if their names are different

**Review by ppolewicz — APPROVED** [review]
